### ISSUE TITLE
Strange pubag IRIs from wikidata

### ISSUE THREAD

**Issue by bigerl** [TortugaAttack/rdf4cpp#308]
Other resources from wikidata causing issues:
```
594157365:136(bad iri): invalid iri. InvalidPath. note: position may not be accurate and instead point to the end of the triple.
601092381:138(bad iri): invalid iri. InvalidPath. note: position may not be accurate and instead point to the end of the triple.
1322066100:138(bad iri): invalid iri. InvalidPath. note: position may not be accurate and instead point to the end of the triple.
```

belonging to these resources:
```
<http://www.wikidata.org/entity/Q63198055> <http://www.wikidata.org/prop/direct-normalized/P7792> <pubag:[%22509826%22,%20%22509826%22]> .
<http://www.wikidata.org/entity/Q64239658> <http://www.wikidata.org/prop/direct-normalized/P7792> <pubag:[%226308582%22,%20%226308582%22]> .
<http://www.wikidata.org/entity/Q28191788> <http://www.wikidata.org/prop/direct-normalized/P7792> <pubag:[%225230079%22,%20%225230079%22]> .
```

These confuse me. %20=` ` and %22=`"`

update: I think I figured it out. Providing an authority (e.g., a domain) in the `hier-part` of an URI (same in the derived IRI specs) is optional.  The URI starts only with `scheme://` if a authority is provided, otherwise, it goes `scheme:path`. See [RFC3986](https://datatracker.ietf.org/doc/html/rfc3986#section-3)). 

Applied to one of the IRIs from wikidata: 
```
  pubag:[%225230079%22,%20%225230079%22]
  └─┬─┘ └──────────────┬───────────────┘
  scheme            path
```

BUT: The characters `[` and `]` are still illegal. 
citing from [RFC3987](https://datatracker.ietf.org/doc/html/rfc3987#section-3.1)
> Systems accepting IRIs MAY also deal with the printable characters in US-ASCII that are not allowed in URIs, namely "<", ">", '"', space, "{", "}", "|", "\", "^", and "`", in step 2 above.  **If these characters are found but are not converted, then the conversion SHOULD fail.  Please note that the number sign ("#"), the percent sign ("%"), and the square bracket characters ("[", "]") are not part of the above list and MUST NOT be converted.**  Protocols and formats that have used earlier definitions of IRIs including these characters **MAY require percent-encoding of these characters as a preprocessing step** to extract the actual IRI from a given field.  This preprocessing MAY also be used by applications allowing the user to enter an IRI.

In conclusion I suggest to follow what the standard says we MAY do: preprocessing the IRI and replace `[` and `]` with percent encoding.

**Comment by AtesComp** [comment]
It's HTML notation.  %20 is space.  %22 is a double quote (").

**Comment by bigerl** [comment]
I think I wasn't clear there. What I am wondering about is how `<pubag:[%225230079%22,%20%225230079%22]>`, or maybe also `<pubag:["25230079", "5230079"]>` are valid IRIs.

edit: figured it out. see update above

**Comment by bigerl** [comment]
Discussion:
- we would need to preprocess explictly for pubag IRIs. That is up to the user but should not be part of rdf4cpp
